proposals/service_api_calls.md

```diff
@@ -0,0 +1,186 @@
+
+Changes for the APICall object:
+
+```sh
+ ❯ kubectl explain cpol.spec.rules.context.apiCall
+
+KIND:     ClusterPolicy
+VERSION:  kyverno.io/v1
+
+RESOURCE: apiCall <Object>
+
+DESCRIPTION:
+     APICall is an HTTP request to the Kubernetes API server, or other JSON web
+     service. The data returned is stored in the context with the name for the
+     context entry.
+
+FIELDS:
+   jmesPath     <string>
+     JMESPath is an optional JSON Match Expression that can be used to transform
+     the JSON response returned from the server. For example a JMESPath of
+     "items | length(@)" applied to the API server response for the URLPath
+     "/apis/apps/v1/deployments" will return the total count of deployments
+     across all namespaces.
+
+   service      <Object>
+     Service is an API call to a JSON web service
+
+   urlPath      <string>
+     URLPath is the URL path to be used in the HTTP GET request to the
+     Kubernetes API server (e.g. "/api/v1/namespaces" or
+     "/apis/apps/v1/deployments"). The format required is the same format used
+     by the `kubectl get --raw` command.
+
+```
+
+New ServiceCall object:
+
+```sh
+❯ kubectl explain cpol.spec.rules.context.apiCall.service
+
+KIND:     ClusterPolicy
+VERSION:  kyverno.io/v1
+
+RESOURCE: service <Object>
+
+DESCRIPTION:
+     Service is an API call to a JSON web service
+
+FIELDS:
+   caBundle     <string>
+     CABundle is a PEM encoded CA bundle which will be used to validate the
+     server certificate.
+
+   data <[]Object>
+     Data specifies the POST data sent to the server.
+
+   requestType  <string> -required-
+     RequestType is the HTTP request type (GET or POST).
+
+   urlPath      <string> -required-
+     URL is the JSON web service URL. The typical format is
+     `https://{service}.{namespace}:{port}/{path}`.
+
+```
+
+
+## Sample Policy Fragments
+
+An HTTP GET API call to a service passing in the `serviceAccountName` and storing the result as `allowed`.
+
+```yaml
+    context:
+    - name: allowed
+        apiCall:
+        service:
+            url: https://service.namespace:443/check?sa={{serviceAccountName}}
+```
+
+An HTTP POST API call to a service that:
+- passes in a map with one entry `images` and value containing a list of images e.g. {"images":["https://ghcr.io/tomcat/tomcat:9","https://ghcr.io/vault/vault:v3","https://ghcr.io/busybox/busybox:latest"]} 
+- uses a X.509 certificate stored in the context variable `cert`
+- applies a JMESPath transformation to check for failed items and stores results in in the context variable `failedCount`
+
+```yaml
+    context:
+    - name: failedCount
+      apiCall:
+        service:
+          url: https://service.namespace:443/validateImages
+          caBundle: {{ cert }}
+          requestType: POST
+          data:
+          - key: images
+            value: "{{ images.[containers, initContainers, ephemeralContainers][].*.reference[] }}"
+        jmesPath: "items[?status == 'failed'] | length(@)"
```

Comment:
I would suggest removing jmespath from the service field. We support variable reassignment and jmespath transformations using inline context variables. We can probably use that to do any transformations in a consistent fashion rather than introducing it here. I know we have this field in other external data calls, but all of them are half baked and miss features that inline variables offer like setting defaults in case values are missing.

Reply:
This is an existing attribute, under APICall. It is used for the Kubernetes API server calls, so was planning to also reuse for any other service call. Its optional.